training/src/main/scala/com/airbnb/aerosolve/training/photon/ml/data/ConvertExamplesToPhotonMLAvroJob.scala

```diff
@@ -0,0 +1,170 @@
+
+    val outputPath = new Path(taskConfig.getString("output_path"))
+
+    val fs = outputPath.getFileSystem(new Configuration())
+    fs.delete(outputPath, true)
+
+    val avroSchemaStr = avroSchema.toString()
+    logger.info(s" Configured output Avro schema: ${avroSchema.toString(true)}")
+
+    val metaDataExtractorClass = Try(taskConfig.getString("meta_data_extractor_class"))
+      .getOrElse(classOf[SingleFloatFamilyMetaExtractor].getName())
+
+    val metaDataExtractorContext = Try(taskConfig
+      .getConfig("meta_data_extractor_context").entrySet().asScala
+      .map { case entry => (entry.getKey(), entry.getValue().toString()) }.toMap
+    ).getOrElse(immutable.Map[String, String]())
+
+    val trainingExampleAvros = transformedData.mapPartitions { case iter =>
+      val schema = new Schema.Parser().parse(avroSchemaStr)
+
+      val metaDataExtractor = Class.forName(metaDataExtractorClass)
+        .newInstance().asInstanceOf[ExampleMetaDataExtractor]
+
+      iter.map { case example =>
+        val record = new GenericData.Record(schema)
+
+        // TODO label format should be more robust
+        val label = if (example.getExample().get(0)
+          .floatFeatures.get(GenericPipeline.LABEL).get("") > 0) {
+          1d
+        } else {
+          0d
+        }
+
+        // TODO decide which key to provide for training example, currently this is not a must-have field
+        // record.put(AVRO_UID, key)
+        record.put(AVRO_LABEL, label)
+        record.put(AVRO_META_DATA_MAP,
+          metaDataExtractor.buildMetaDataMap(example, metaDataExtractorContext))
+
+        createAvroFeatureArrays(record, example, schema)
+
+        record
+      }
+    }
+
+    val job = Job.getInstance
+    AvroJob.setOutputKeySchema(job, avroSchema)
+    val hadoopConf = job.getConfiguration()
+
+    // Compress Avro binary files
+    hadoopConf.set("mapreduce.output.fileoutputformat.compress", "true")
+    hadoopConf.set(AvroJob.CONF_OUTPUT_CODEC, "deflate")
+
+    // Save as Avro records
+    trainingExampleAvros
+      .map { r => (new AvroKey(r), NullWritable.get()) }
+      .saveAsNewAPIHadoopFile(outputPath.toString(),
+        classOf[AvroKey[GenericRecord]],
+        classOf[org.apache.hadoop.io.NullWritable],
+        classOf[AvroKeyOutputFormat[GenericRecord]],
+        hadoopConf)
+  }
+
+  // Load RDD[Example] according to context information
+  private def loadExamples(sc: SparkContext,
+                           taskConfig: Config): (RDD[Example], Schema) = {
+
+    val hiveTraining = GenericPipeline.runQuery(sc, taskConfig.getString("input_hive_query"))
+    val dataFrameSchema = hiveTraining.schema.fields
+
+    val namespaces = (dataFrameSchema.map { case f =>
+      val idx = f.name.indexOf("_")
+
+      if (idx == -1) {
+        f.name
+      } else {
+        f.name.substring(0, idx)
+      }
+    }.toSet[String] ++ taskConfig.getStringList("avro_extra_namespaces").asScala).toArray
```

Comment:
Any extra family names generated by transformers but not presented in raw Hive inputs should be added in `avro_extra_namespaces`. We could also call the transformer to reflect such names but I figured this way is probably easier since we are gonna declare those transformers needed in the end anyway.